content/advent-2015/composable-pipelines-improvements.md

```diff
@@ -0,0 +1,268 @@
+proc2.In = proc1.OutChan()
+```
+
+## A better way
+
+What I realized, is that if we use a function to create new tasks, that pre-populates all the channel fields, we can
+just assign those fields to each other.
+
+So, a task would be defined like so, including its factory function:
+
+```go
+type AProcess struct {
+	In  string
+	Out string
+}
+
+func NewProcess(*AProcess) {
+	return &AProcess{
+		In:  make(chan string),
+		Out: make(chan string),
+	}
+}
+
+func (p *AProcess) Init() {
+	go func() {
+		for line := range p.In {
+			// Do something with line ...
+		}
+	}()
+}
+```
+
+Then, to connect two such processes together, we would go:
+
+```go
+proc1 := NewProcess()
+proc2 := NewProcess()
+// Connect
+proc2.In = proc1.Out
+```
+... and we could even connect the other way around, since both the In- and Out-port fields are initiated with channels,
+so it doesn't matter which of these channels we use, as long as it is the same channel used on the corresponding out-
+and in-port:
+
+```go
+// <snip>
+// Connect the other way around, setting the out-port to the channel in the in-port
+proc1.Out = proc2.In
+```
+
+## Let's create a pipeline component too
+
+There was one other ugly part of that previously blogged example. In order to drive the execution of a set of connected
+processes, we were just looping over the output of the out-port of the last component, right in the programs main-method.
+
+That is, the following part in the previous post:
+
+```go
+for i := range printer.DrivingBeltChan() {
+	linesProcessed += i
+}
+```
+
+This was due to the fact that (as far as I know) the execution of separate go-routines can not really be started until they get
+a signal from the main go-routine, over a channel, for example.
+
+I couldn't come up with a better suggestion for how to drive such a chain of processes, until [Egon Elbre](https://twitter.com/egonelbre)
+[elaborated on some tips](https://groups.google.com/forum/#!msg/golang-nuts/vgj_d-MjUHA/T9sE64Yrcq0J) on how to enhance the pattern.
+
+While I did not use Egon's whole suggestion since it included a fair bit of reflection and departed from my idea
+of a framework-less pattern, his code examples did a nice trick; Rather than letting the processes fire up go-routines,
+like my pattern did inside the `Init()` methods, he had plain `Run()` methods without any go-routines in them, and instead
+fired off as go-routines (with the `go` keyword) outside of the processes.
+
+So, if we replaced the `Init()` method in the code examples above with the following `Run()` method:
+
+```go
+func (p *AProcess) Run() {
+	for line := range p.In {
+		// Do something with line ...
+	}
+}
+```
+
+... then we would execute these `Run()` methods like so (adapting it to our example code here):
+
+```go
+proc1 := NewProcess()
+proc2 := NewProcess()
+go proc1.Run()
+go proc2.Run()
+```
+
+This suggests an elegant solution of the problem of driving a chain of go-routines from the main thread: Just skip the go
```

Comment:
A couple suggests to make this sentence easier to read.

Change:
> elegant solution **of** the problem

to

> elegant solution **to** the problem

and change:
> thread: Just skip the

to

> thread. Skip the